assets/js/wfmodule/WfModule.test.js

```diff
@@ -46,44 +64,83 @@ describe('WfModule, not read-only mode', () => {
       'onParamChanged' : okResponseMock()
     };
 
-    props = {
-        isReadOnly: false,
-        name: 'TestModule',
-        wfModule: wf_module,
-        changeParam: mockApi.onParamChanged,
-        removeModule:  () => {} ,
-        revision: 707,
-        selected: true,
-        api: mockApi,
-        connectDragSource:  jest.fn(),
-        connectDropTarget:  jest.fn(),
-        connectDragPreview: jest.fn(),
-        focusModule:        jest.fn(),
-        startDrag :         jest.fn(),
-        stopDrag :          jest.fn()
-      };
-
-    wrapper = mount(
-      <WfModule
-        {...props}
-      />
-    )
-  });
-
-  it('Renders and calls functions on componentDidMount and render', () => {
+    props = buildShallowProps()
+  })
+
+  it('matches snapshot', () => {
+    const wrapper = shallow(<WfModule {...props}/>)
     expect(wrapper).toMatchSnapshot();
-    expect(props.connectDragSource.mock.calls.length).toBe(1);
-    expect(props.connectDropTarget.mock.calls.length).toBe(1);
-    expect(props.focusModule.mock.calls.length).toBe(1);
-  });
-
-  it('getParamText and setParamText', () => {
-    expect(wrapper.instance().getParamText('url')).toBe(wf_module.parameter_vals[0].value);
-
-    wrapper.instance().setParamText('url','http://foocastle.ai');
-    expect(props.changeParam.mock.calls.length).toBe(1);
-    expect(props.changeParam.mock.calls[0][0]).toEqual(wf_module.parameter_vals[0].id);  // should translate id_name to id
-    expect(props.changeParam.mock.calls[0][1]).toEqual({'value' : 'http://foocastle.ai'});
   })
 
-});
+  it('is draggable', () => {
+    shallow(<WfModule {...props}/>)
+    expect(props.connectDragSource).toHaveBeenCalled()
+    expect(props.connectDropTarget).toHaveBeenCalled()
+  })
+
+  it('auto-focuses if selected', () => {
+    shallow(<WfModule {...props} selected={true}/>)
+    expect(props.focusModule).toHaveBeenCalled()
+  })
+
+  it('does not auto-focus if not selected', () => {
+    shallow(<WfModule {...props} selected={false}/>)
+    expect(props.focusModule).not.toHaveBeenCalled()
+  })
+
+  it('supplies getParamText and setParamText', () => {
+    const wrapper = shallow(<WfModule {...props}/>)
+    const instance = wrapper.instance()
+
+    expect(instance.getParamText('url')).toEqual('http://some.URL.me')
+    wrapper.instance().setParamText('url', 'http://foocastle.ai');
+
+    expect(props.changeParam).toHaveBeenCalledWith(100, { value: 'http://foocastle.ai' })
+    // and the input prop should not be mutated
+    expect(wf_module.parameter_vals[0].value).toEqual('http://some.URL.me')
+  })
+
+  describe('mapStateToProps', () => {
```

Comment:
I agree: both "verbose" and "underpowered." Not quite sure what to do about it, though....:

* In general, `mount()` will reduce "underpowered" but double "verbose" and increase fragility. ("Fragility" = "likelihood you'll break the test when you don't break the code") I used `mount()` in [ModuleSearch.test.js](https://github.com/CJWorkbench/cjworkbench/blob/master/assets/js/ModuleSearch.test.js#L98-L120). There, it cost 22 lines of fragile boilerplate setup code; furthermore, testing for matches (`.toBe(true)`/`.toBe(false)` became much larger: 5 big lines per test instead of 3. If we were to tighten these as much as we could, we'd get 40 lines of boilerplate and 2 lines per test.
* `WfModule.test.js` in particular would be very fragile to changes in the DOM structure ... which we desperately need.
* I used test-driven development to write `mapStateToProps()`. I think the function is complex enough to warrant its own tests, and I had a good TDD experience when I wrote it. So maybe the best approach is to write all these tests I wrote _and_ write another test per property to make sure a state change leads to a render?

I assume you mean for this comment to be part of a larger discussion about how we should test? I hear that and agree. I tried one strategy with `ModuleSearch.test.js` and another with `WfModule.test.js`. I'm unsatisfied with both. Perhaps next time we should try a hybrid?

Reply:
Also:

Next time, maybe we should check the `lesson_highlight` JSON in the test suite, using `LessonHighlightsType` (a `PropType`). I had typo problems because I wrote `name:` instead of `moduleName:`, etc. The test suite should correct me. And/or maybe we should write factories just for our test suite. They'd be particularly helpful to build `store.getState().workflow`.